feat(frontend): Implement delegated moderation UI

tripledoublev opened the pull request:
UI for delegated moderation: 
- allows streamers to grant moderation permissions and delegated moderators to perform moderation actions.

Changes:                                                 
- **ModeratorPanel** component for permission management       
- **Moderation hooks**: `useCanModerate`, `useModeratorManagement`, `useCreateBlockRecord`, etc.         
- **Actions**: block users, hide messages, update livestream metadata                                     
- **Real-time updates**: WebSocket integration for permission sync                                                
                                                          
Tested in browser with moderation permissions, delegated actions, and WebSocket updates.  


Built on top of #788 
                                 
Part 4 of 4 for #718

Closes #718

tripledoublev committed:
feat(frontend): Add moderation permission hooks and moderator management

- Add useCanModerate hook to check moderation permissions
  * Reads from livestream store (updated via WebSocket)
  * Supports owner and delegated moderator permissions
  * Handles permission expiration

- Add moderator management hooks
  * useListModerators - list all moderators for current user
  * useAddModerator - add new moderator with permissions
  * useRemoveModerator - remove moderator by rkey
  * Resolves handles to DIDs automatically

tripledoublev committed:
feat(frontend): Add ModeratorPanel UI component

- Add ModeratorPanel component for managing moderators
  * List of moderators with permissions and expiration
  * Add moderator dialog with permission selection
  * Remove moderator with confirmation
  * Resolves DIDs to handles for display
  * Supports embedded mode for use in other panels

tripledoublev committed:
feat(frontend): Integrate moderation permissions with livestream store

- Add moderationPermissions to livestream state
  * Stores moderation permission records from streamer's repo
  * Updated in real-time via WebSocket

- Handle moderation permission records in WebSocket consumer
  * Add/update permissions when records are created
  * Clear and refetch when records are deleted
  * Enables real-time permission updates without page refresh

tripledoublev committed:
feat(frontend): Update chat moderation to use delegated moderation

- Update block and hide hooks to support delegated moderation
  * Pass streamerDID to moderation actions
  * Use XRPC endpoints when caller is not the streamer
  * Direct ATProto writes when caller is the streamer

- Update ModView to use useCanModerate hook
  * Replace useIsMyStream with permission-based checks
  * Show moderation actions based on canHide/canBan permissions
  * Support both streamer and delegated moderators

- Cleanup: Remove unused moderationSvcDid parameter

tripledoublev committed:
feat(frontend): Integrate ModeratorPanel into stream settings

- Add "Moderations" tab to Stream Settings panel
- Embed ModeratorPanel as sub-panel for cleaner UI
- Accessible via button selector alongside Create and Metadata

tripledoublev committed:
feat(frontend): Support multiple delegation records per moderator

Update frontend state management to handle multiple separate delegation
records per moderator, matching backend bug fix.

- Merge permissions from all delegation records in useCanModerate
- Fix WebSocket record matching using rkey + createdAt
- Skip expired delegations during permission merge

Related to #718

tripledoublev committed:
feat(frontend): Add livestream title management to delegated moderation

Enable delegated moderators with livestream.manage permission to update
stream titles via mod view.

- Add useUpdateLivestreamRecord hook with owner/delegated paths
- Implement update stream title dialog in mod view
- Add livestream management permission checks
- Export block module from streamplace-store

Note: block.tsx now contains multiple moderation actions.

Related to #718

espeon commented on js/components/src/streamplace-store/moderator-management.tsx:
```diff
@@ -0,0 +1,180 @@
+          throw new Error(
+            `Invalid DID or handle: ${moderatorDID}. Please use a valid DID (did:plc:...) or handle (@handle.bsky.social)`,
+          );
+        }
```
Could do this more simply IMO

Something like:
```suggestion
      if (!moderatorDID.startsWith("did:")) {
        if (moderatorDID.startsWith("@")) {
          moderatorDID = moderatorDID.substring(1), // Remove @
        } 
        try {
          const resolved = await agent.com.atproto.identity.resolveHandle({
            handle: moderatorDID,
          });
          moderatorDID = resolved.data.did;
        } catch (e) {
          // If resolution fails, assume it's already a DID
          throw new Error(
            `Invalid DID or handle: ${moderatorDID}. Please use a valid DID (did:plc:...) or handle (@handle.bsky.social)`,
          );
        }
```
and then we fall back to the did. Later I'd like to hook this up with the autofill API I have in another PR.

espeon commented on js/components/src/components/chat/mod-view.tsx:
```diff
@@ -78,125 +95,262 @@ export const ModView = forwardRef<ModViewRef, ModViewProps>(() => {
+      </DropdownMenu>
+
+      {/* Update Stream Title Dialog - rendered outside dropdown */}
+      {showUpdateTitleDialog && (
```
Not 100% sure this should be in the mod view, but not sure where else to put it. I'd either put it in video settings, or make some sort of quick stream settings accessable near the share menu?

espeon commented on js/components/src/components/dashboard/moderator-panel.tsx:
```diff
@@ -0,0 +1,625 @@
+            borders.color.red[700],
+            mb[4],
+          ]}
+        >
+          <Text style={[textStyle.red[400], { fontSize: 13 }]}>{error}</Text>
+        </View>
+      )}
+
+      {/* Actions */}
+      <DialogFooter>
+        <Button variant="secondary" onPress={onClose} disabled={isLoading}>
+          <Text>Cancel</Text>
+        </Button>
+        <Button
+          variant="primary"
+          onPress={handleAdd}
+          disabled={
+            isLoading ||
+            !moderatorDID.trim() ||
+            Object.values(permissions).every((p) => !p)
+          }
+        >
+          <Text>{isLoading ? "Adding..." : "Add Moderator"}</Text>
+        </Button>
+      </DialogFooter>
```
My fault for changing this on y'all but this should be

```suggestion
      <DialogFooter>
        <Button
          width="min"
          variant="secondary"
          onPress={onClose}
          disabled={isLoading}
        >
          Cancel
        </Button>
        <Button
          width="min"
          variant="primary"
          onPress={handleAdd}
          disabled={
            isLoading ||
            !moderatorDID.trim() ||
            Object.values(permissions).every((p) => !p)
          }
        >
          {isLoading ? "Adding..." : "Add Moderator"}
        </Button>
      </DialogFooter>
```

WIll try to find a better solution, I know this is a horrible default.

espeon commented on js/components/src/streamplace-store/moderation.tsx:
```diff
@@ -0,0 +1,185 @@
+
+      try {
+        // Use authenticated agent to list permission records from the streamer's repo
+        const result = await agent.com.atproto.repo.listRecords({
+          repo: streamerDID,
+          collection: "place.stream.moderation.permission",
+          limit: 100,
+        });
```
This works but IMO it would be nice to have an xrpc endpoint to get these on the server side - we should have them anyways(?) so it saves us a round trip to the PDS.

espeon commented on js/components/src/components/chat/mod-view.tsx:
```diff
@@ -277,3 +433,159 @@ export function ReportButton({
+            </Text>
+          </View>
+        )}
+      </View>
+
+      <DialogFooter>
+        <Button
+          variant="secondary"
+          onPress={() => {
+            onClose();
+            setError(null);
+            setTitle(livestream?.record?.title || "");
+          }}
+          disabled={isLoading}
+        >
+          <Text>Cancel</Text>
+        </Button>
+        <Button
+          variant="primary"
+          onPress={handleUpdate}
+          disabled={isLoading || !title.trim()}
+        >
+          <Text>{isLoading ? "Updating..." : "Update Title"}</Text>
+        </Button>
+      </DialogFooter>
```
Got into it a bit more in another comment but this should be changed to
```suggestion
      <DialogFooter>
        <Button
          width="min"
          variant="secondary"
          onPress={() => {
            onClose();
            setError(null);
            setTitle(livestream?.record?.title || "");
          }}
          disabled={isLoading}
        >
          <Text>Cancel</Text>
        </Button>
        <Button
          variant="primary"
          width="min"
          onPress={handleUpdate}
          disabled={isLoading || !title.trim()}
        >
          <Text>{isLoading ? "Updating..." : "Update Title"}</Text>
        </Button>
      </DialogFooter>
```

espeon reviewed (changes requested):
Seems good so far! I have a few nitpicks and one bigger question about where the change stream title should go.

tripledoublev committed:
fix: address review feedback from espeon

- Add width="min" to dialog buttons for proper layout
- Replace text checkmark with Check icon from lucide-react-native
- Fix typo: mb[4] → my[4] in moderator panel
- Simplify DID resolution logic with cleaner nested structure
- Change livestream update to CREATE for chapter marker pattern

Addresses feedback from PR #796

makew0rld committed:
make fix
